source/modulo_core/include/modulo_core/concepts.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <geometry_msgs/msg/accel_stamped.hpp>
+#include <geometry_msgs/msg/pose_stamped.hpp>
+#include <geometry_msgs/msg/transform_stamped.hpp>
+#include <geometry_msgs/msg/twist_stamped.hpp>
+#include <geometry_msgs/msg/wrench_stamped.hpp>
+#include <sensor_msgs/msg/joint_state.hpp>
+#include <std_msgs/msg/bool.hpp>
+#include <std_msgs/msg/float64.hpp>
+#include <std_msgs/msg/float64_multi_array.hpp>
+#include <std_msgs/msg/int32.hpp>
+#include <std_msgs/msg/string.hpp>
+#include <tf2_msgs/msg/tf_message.hpp>
+
+#include "modulo_core/EncodedState.hpp"
+
+#include <state_representation/State.hpp>
+#include <state_representation/space/cartesian/CartesianState.hpp>
+#include <state_representation/space/joint/JointState.hpp>
+
+namespace modulo_core::concepts {
+
+// Data type concepts
+
+template<typename DataT>
+concept StateRepresentationDataT = std::derived_from<DataT, state_representation::State>;
+
+template<typename DataT>
+concept CorePrimitivesT = std::same_as<DataT, int> || std::same_as<DataT, float> || std::same_as<DataT, double>
+    || std::same_as<DataT, bool> || std::same_as<DataT, std::string> || std::same_as<DataT, std::vector<double>>;
+
+template<typename DataT>
+concept CoreDataT = StateRepresentationDataT<DataT> || CorePrimitivesT<DataT>;
+
+template<typename DataT>
+concept CustomDataT = !CoreDataT<DataT>;
+
+// Translation concepts
+
+template<typename DataT>
+concept TranslatedDataT = std::same_as<DataT, geometry_msgs::msg::Accel>
+    || std::same_as<DataT, geometry_msgs::msg::AccelStamped> || std::same_as<DataT, geometry_msgs::msg::Pose>
+    || std::same_as<DataT, geometry_msgs::msg::PoseStamped> || std::same_as<DataT, geometry_msgs::msg::Transform>
+    || std::same_as<DataT, geometry_msgs::msg::TransformStamped> || std::same_as<DataT, geometry_msgs::msg::Twist>
+    || std::same_as<DataT, geometry_msgs::msg::TwistStamped> || std::same_as<DataT, geometry_msgs::msg::Wrench>
+    || std::same_as<DataT, geometry_msgs::msg::WrenchStamped> || std::same_as<DataT, sensor_msgs::msg::JointState>
+    || std::same_as<DataT, tf2_msgs::msg::TFMessage> || std::same_as<DataT, std_msgs::msg::Bool>
+    || std::same_as<DataT, std_msgs::msg::Float64> || std::same_as<DataT, std_msgs::msg::Float64MultiArray>
+    || std::same_as<DataT, std_msgs::msg::Int32> || std::same_as<DataT, std_msgs::msg::String>;
+
+template<typename DataT>
```

Comment:
I guess the `TranslatedOrEncodedDataT` and `NonTranslatedDataT` will be required for the integration in the components?

Also, why not add `|| std::same_as<DataT, EncodedState>` directly to `TranslatedDataT`?

Reply:
I will now even go as far as saying that the unused concepts could be removed from this PR such that it doesn't go out of scope

Reply:
I do need to type check separately, but I do agree that this is a bit out of scope with the new form of this PR. I will remove it now and then we can discuss again in the corresponding PR